src/Microsoft.AspNetCore.OData/Query/Query/DefaultSkipTokenHandler.cs

```diff
@@ -228,234 +300,154 @@ internal static string GenerateSkipTokenValue(Object lastMember, IEdmModel model
-                if (orderBy != null)
-                {
-                    orderByNodes = orderBy.OrderByNodes;
-                }
-            }
-
-            return ApplyToCore(query, querySettings, orderByNodes, skipTokenQueryOption.Context, skipTokenQueryOption.RawValue);
+            return ApplyToCore(query, querySettings, queryOptions, skipTokenQueryOption.RawValue);
         }
 
-        /// <summary>
-        /// Core logic for applying the query option to the IQueryable. 
-        /// </summary>
-        /// <param name="query">The original <see cref="IQueryable"/>.</param>
-        /// <param name="querySettings">Query setting used for validating the query option.</param>
-        /// <param name="orderByNodes">OrderBy information required to correctly apply the query option for default implementation.</param>
-        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
-        /// <param name="skipTokenRawValue">The raw string value of the skiptoken query parameter.</param>
-        /// <returns></returns>
-        [SuppressMessage("Microsoft.Maintainability", "CA1506:AvoidExcessiveClassCoupling", Justification = "Class coupling acceptable.")]
-        private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes, ODataQueryContext context, string skipTokenRawValue)
+        private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings querySettings,
+             ODataQueryOptions queryOptions, string skipTokenRawValue)
         {
-            Contract.Assert(query != null);
-            Contract.Assert(context.ElementClrType != null);
-
-            IDictionary<string, OrderByDirection> directionMap;
-            if (orderByNodes != null)
+            OrderByQueryOption orderByOption = queryOptions?.GenerateStableOrder();
+            if (orderByOption == null)
             {
-                directionMap =
-                    orderByNodes.OfType<OrderByPropertyNode>().ToDictionary(node => context.Model.GetClrPropertyName(node.Property), node => node.Direction);
-            }
-            else
-            {
-                directionMap = new Dictionary<string, OrderByDirection>();
+                return query;
             }
 
-            IDictionary<string, (object PropertyValue, Type PropertyType)> propertyValuePairs = PopulatePropertyValuePairs(skipTokenRawValue, context);
+            // It's better to visit the nodes of OrderByClause reclusively to get the orderby raw value.
+            // Since we don't have such methods, let's simply split the request raw value.
+            IList<string> orderBys = orderByOption.RawValue.Split(',');
+            IList<OrderByClause> orderByClauses = orderByOption.OrderByClause.ToList();
+            IList<string> tokenValueParis = PopulatePropertyValuePairs(skipTokenRawValue);
```

Comment:
```suggestion
            IList<string> tokenValuePairs = PopulatePropertyValuePairs(skipTokenRawValue);
```

Reply:
Actually, since this list only contains the values (without keys), why not just `tokenValues`?

```suggestion
            List<string> tokenValues = PopulatePropertyValuePairs(skipTokenRawValue);
```